pandas/core/internals/blocks.py

```diff
@@ -1174,6 +1177,9 @@ def coerce_to_target_dtype(self, other):
                 f"possible recursion in coerce_to_target_dtype: {self} {other}"
             )
 
+        if is_categorical_dtype(dtype) or self.is_datetime:
```

Comment:
the is_datetime condition is already handled above, if for some reason you actually need to do this then do it above (and I am not really sure why you are trying to do this)